drivers/include/periph/gpio_ll.h

```diff
@@ -0,0 +1,628 @@
+ * @author      Marian Buschsieweke <marian.buschsieweke@ovgu.de>
+ *
+ * @warning     This API is not stable yet and intended for internal use only as
+ *              of now.
+ */
+
+#ifndef PERIPH_GPIO_LL_H
+#define PERIPH_GPIO_LL_H
+
+#include <inttypes.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "architecture.h"
+#include "periph_cpu.h"
+#include "periph/gpio.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   GPIO port type
+ */
+typedef uintptr_t gpio_port_t;
+
+#if !defined(GPIO_PORT_UNDEF) || defined(DOXYGEN)
+/**
+ * @brief   Magic "undefined GPIO port" value
+ */
+#define GPIO_PORT_UNDEF         UINTPTR_MAX
+#endif
+
+#ifdef DOXYGEN
+/**
+ * @brief   Get the @ref gpio_port_t value of the port identified by @p num
+ *
+ * @note    If @p num is a compile time constant, this is guaranteed to be
+ *          suitable for a constant initializer.
+ *
+ * Typically this will be something like `(GPIO_BASE_ADDR + num * sizeof(struct
+ * vendor_gpio_reg))`
+ */
+#define GPIO_PORT(num)  implementation_specific
+#endif
+
+#ifdef DOXYGEN
+/**
+ * @brief   Get the number of the GPIO port belonging to the given @ref
+ *          gpio_port_t value
+ *
+ * @note    If @p port is a compile time constant, this is guaranteed to be
+ *          suitable for a constant initializer.
+ *
+ * @pre     @p port is the return value of @ref GPIO_PORT
+ *
+ * For every supported port number *n* the following `assert()` must not blow
+ * up:
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.c}
+ * assert(n == GPIO_PORT_NUM(GPIO_PORT(n)));
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ */
+#define GPIO_PORT_NUM(port) implementation_specific
+#endif
+
+#if !defined(HAVE_GPIO_STATE_T) || defined(DOXYGEN)
+/**
+ * @brief   Enumeration of GPIO states (direction)
+ */
+typedef enum {
+    /**
+     * @brief   Use pin as output in push-pull configuration
+     *
+     * | Logical Value  | Electrical Behavior               |
+     * |:-------------- |:--------------------------------- |
+     * | `0`            | Low                               |
+     * | `1`            | High                              |
+     */
+    GPIO_OUTPUT_PUSH_PULL,
+    /**
+     * @brief   Use pin as output in open collector configuration
+     *
+     * | Logical Value  | Electrical Behavior               |
+     * |:-------------- |:--------------------------------- |
+     * | `0`            | Low                               |
+     * | `1`            | High Impedance (Disconnected)     |
+     */
+    GPIO_OUTPUT_OPEN_DRAIN,
+    /**
+     * @brief   Use pin as output in open emitter configuration
+     *
+     * | Logical Value  | Electrical Behavior               |
+     * |:-------------- |:--------------------------------- |
+     * | `0`            | High Impedance (Disconnected)     |
+     * | `1`            | High                              |
+     */
+    GPIO_OUTPUT_OPEN_EMITTER,
```

Comment:
I changed it to source. As far as I know only term (and feature) Open Drain is common. But since the NRF5* also exposes Open Source behavior, I thought it would be a shame to prevent using it.